# 本周大綱
+ 自動更新 header
+ 網頁編碼
+ json
+ post 請求
+ API
+ 3 個爬蟲模組
+ pandas 爬蟲
+ 下載 chromedriver

不用抄，下課之前會給這份檔案

## 自動更新 header
瀏覽器每隔一段時間就會自動更新版本，雖然不更新 header 也是可以跑，但因為瀏覽器會自動更新，理論上不應該存在太舊的 chrome 瀏覽器版本。想要裝得像，一勞永逸的方法還是自動更新爬蟲的 header

In [1]:
from os import listdir # 列出指定路徑的檔案清單

listdir(r'C:\Program Files (x86)\Google\Chrome\Application')

['136.0.7103.114',
 'chrome.exe',
 'chrome.VisualElementsManifest.xml',
 'chrome_proxy.exe',
 'debug.log',
 'Dictionaries',
 'master_preferences',
 'SetupMetrics']

In [2]:
chrome_version = listdir(r'C:\Program Files (x86)\Google\Chrome\Application')[0].split('.')[0]
print(chrome_version)

136


In [10]:
from os import listdir

chrome_version = listdir(r'C:\Program Files (x86)\Google\Chrome\Application')[0].split('.')[0]
header = {'User-Agent': f'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/{chrome_version}.0.0.0 Safari/537.36'}
print(header)

{'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/136.0.0.0 Safari/537.36'}


## 網頁編碼
每個網頁都會設定編碼方式，requests 預設網頁的編碼方式是 UTF8，中文網站比較可能是 UTF8 或是 BIG5，有時候抓到的網頁原始碼是亂碼，就是因為編碼方式不對。例如抓台灣證券交易所的最新訊息 https://www.twse.com.tw/zh/products/securities/etf/news.html

In [1]:
from os import listdir
import requests
from bs4 import BeautifulSoup

chrome_version = listdir(r'C:\Program Files (x86)\Google\Chrome\Application')[0].split('.')[0]
header = {'User-Agent': f'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/{chrome_version}.0.0.0 Safari/537.36'}

url = 'https://www.twse.com.tw/zh/products/securities/etf/news.html'
r = requests.get(url, headers=header)
soup = BeautifulSoup(r.text, 'html.parser')
soup


<!DOCTYPE html>

<html lang="zh-Hant-tw">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<title>ETF 最新訊息 - TWSE 臺灣證券交易所</title>
<link href="/favicon.ico" rel="shortcut icon"/>
<meta content="臺灣證券交易所å
¨çƒè³‡è¨Šç¶²ä»‹ç´¹å
¬å¸çµ„織、沿革外，並分為「交易資訊」、「上市å
¬å¸ã€ã€ 「產品與服務」、「結算服務」、「市場å
¬å‘Šã€ã€ã€Œæ³•ä»¤è¦ç« ã€ã€ã€ŒæŠ•è³‡äººæ•™è‚²ã€ã€ã€Œçµ±è¨ˆå ±è¡¨ã€ç­‰ï¼ŒæœŸèƒ½æä¾›æŠ•è³‡å¤§çœ¾å³æ™‚完整之訊息，使投資大眾對本å
¬å¸æ¥­å‹™åŠå¸‚å ´å‹•æ
‹æœ‰æ›´æ·±ä¸€å±¤çš„瞭解。" name="description"/>
<meta content="證券,上市,大盤,國際股市指數,匯率,個股行æƒ
等,年報,投資,期貨,台灣50,ETF" name="keywords"/>
<meta content="TWSE" name="author"/>
<link href="/res/css/web.css" rel="stylesheet"/>
<

In [2]:
# 會發現標籤部分正常，但內容部分是亂碼，雖然可以看到在<head>中<meta charset="utf-8"/>表示網頁設定的編碼方式是 UTF8，
# 但基於某些神秘的原因解讀出來是亂碼。這時候只要再指定一次網頁的編碼方式是 UTF8 就可以解決了。

from os import listdir
import requests
from bs4 import BeautifulSoup

chrome_version = listdir(r'C:\Program Files (x86)\Google\Chrome\Application')[0].split('.')[0]
header = {'User-Agent': f'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/{chrome_version}.0.0.0 Safari/537.36'}

url = 'https://www.twse.com.tw/zh/products/securities/etf/news.html'
r = requests.get(url, headers=header)
r.encoding = 'UTF-8' # 指定編碼方式，utf-8、UTF-8、UTF8、utf8 都一樣
soup = BeautifulSoup(r.text, 'html.parser')
soup

# 但沒有抓到資料

<!DOCTYPE html>

<html lang="zh-Hant-tw">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<title>ETF 最新訊息 - TWSE 臺灣證券交易所</title>
<link href="/favicon.ico" rel="shortcut icon"/>
<meta content="臺灣證券交易所全球資訊網介紹公司組織、沿革外，並分為「交易資訊」、「上市公司」、 「產品與服務」、「結算服務」、「市場公告」、「法令規章」、「投資人教育」、「統計報表」等，期能提供投資大眾即時完整之訊息，使投資大眾對本公司業務及市場動態有更深一層的瞭解。" name="description"/>
<meta content="證券,上市,大盤,國際股市指數,匯率,個股行情等,年報,投資,期貨,台灣50,ETF" name="keywords"/>
<meta content="TWSE" name="author"/>
<link href="/res/css/web.css" rel="stylesheet"/>
<meta content="web" name="layout">
<meta content="etf" name="sidebar"/>
<link href="/res/css/web-report.css" rel="stylesheet"/>
<style>
    #links{min-height:auto;margin-bottom:40px}.rwd-table a{color:#000;text-decoration:none}.rwd-table a:hover{text-decoration:underline}.rwd-table th{min-width:auto;max-width:none}.rwd-table th

## 開發人員工具簡介2
開發人員工具第二常用到的是網路(Network)，剛打開的時候這邊會是空的，畫面上會提示要按 Ctrl + R 重新載入才會看到東西。這代表網路區塊只會呈現打開開發人員工具之後，伺服器傳送過來的東西。如果想要看到伺服器傳送過來的所有東西，有兩種方法：
1. 打開開發人員工具之後重新整理
2. 打開網頁之後立刻按下 F12

以台灣證券交易所的最新訊息為例，可以發現網站伺服器不會只送一個東西過來，實際上我們看到的網頁內容是瀏覽器處理完所有這些東西之後才有的樣子。

## json 格式
在剛剛的例子中，雖然不是亂碼了，但也沒有抓到資料。原因在於，伺服器先回傳網頁，格子裡的資料是之後才補上，而 requests 只接收第一次回傳的資料，因此只要找到實際上請求資料的那個網址就好了。步驟如下：
開發人員工具→網路(Network)→Ctrl+R→Fetch/XHR→一個一個點開看預覽→找到想要的資料→標頭→複製請求網址


In [3]:
# 找到之後貼到 url
#url = 'https://www.twse.com.tw/rwd/zh/ETFNews/news?response=json&_=1716956811498'
url = 'https://www.twse.com.tw/rwd/zh/ETFNews/news?response=json' # 刪掉流水號也可以
r = requests.get(url, headers=header)
soup = BeautifulSoup(r.text, 'html.parser')
soup


{"status":"ok","data":[["2025-05-23T05:50:10Z","永豐ESG低碳高息(00930) 分配收益資訊-公告日期 114年05月23日","https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&amp;firstin=1&amp;TYPEK=all&amp;SKIND=F&amp;SEQ_NO=2&amp;DATE1=20250523&amp;COMP=A00025",null],["2025-05-23T05:31:48Z","元大台灣價值高息(00940) 分配收益資訊-公告日期 114年05月23日","https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&amp;firstin=1&amp;TYPEK=all&amp;SKIND=F&amp;SEQ_NO=1&amp;DATE1=20250523&amp;COMP=A00005",null],["2025-05-20T01:07:20Z","元大US高息特別股(00771) 分配收益資訊-公告日期 114年05月20日","https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&amp;firstin=1&amp;TYPEK=all&amp;SKIND=F&amp;SEQ_NO=2&amp;DATE1=20250520&amp;COMP=A00005",null],["2025-05-20T00:54:00Z","元大臺灣ESG永續(00850) 分配收益資訊-公告日期 114年05月20日","https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&amp;firstin=1&amp;TYPEK=all&amp;SKIND=F&amp;SEQ_NO=1&amp;DATE1=20250520&amp;COMP=A00005",null],["2025-05-19T07:22:58Z","群益優選非投等債(00953B) 分配收益資訊-公告日期 114年05月19日","https://mopsov.twse.com.tw/mops/web

In [4]:
# 只要看起來是字典格式，就表示這是 json 格式，所以用 json 模組轉換成字典：
import json
data = json.loads(r.text)
data

{'status': 'ok',
 'data': [['2025-05-23T05:50:10Z',
   '永豐ESG低碳高息(00930) 分配收益資訊-公告日期 114年05月23日',
   'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=2&DATE1=20250523&COMP=A00025',
   None],
  ['2025-05-23T05:31:48Z',
   '元大台灣價值高息(00940) 分配收益資訊-公告日期 114年05月23日',
   'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=1&DATE1=20250523&COMP=A00005',
   None],
  ['2025-05-20T01:07:20Z',
   '元大US高息特別股(00771) 分配收益資訊-公告日期 114年05月20日',
   'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=2&DATE1=20250520&COMP=A00005',
   None],
  ['2025-05-20T00:54:00Z',
   '元大臺灣ESG永續(00850) 分配收益資訊-公告日期 114年05月20日',
   'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=1&DATE1=20250520&COMP=A00005',
   None],
  ['2025-05-19T07:22:58Z',
   '群益優選非投等債(00953B) 分配收益資訊-公告日期 114年05月19日',
   'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&fi

In [5]:
# 取得最新消息
for i in data['data']:
    print(i)

['2025-05-23T05:50:10Z', '永豐ESG低碳高息(00930) 分配收益資訊-公告日期 114年05月23日', 'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=2&DATE1=20250523&COMP=A00025', None]
['2025-05-23T05:31:48Z', '元大台灣價值高息(00940) 分配收益資訊-公告日期 114年05月23日', 'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=1&DATE1=20250523&COMP=A00005', None]
['2025-05-20T01:07:20Z', '元大US高息特別股(00771) 分配收益資訊-公告日期 114年05月20日', 'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=2&DATE1=20250520&COMP=A00005', None]
['2025-05-20T00:54:00Z', '元大臺灣ESG永續(00850) 分配收益資訊-公告日期 114年05月20日', 'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=1&DATE1=20250520&COMP=A00005', None]
['2025-05-19T07:22:58Z', '群益優選非投等債(00953B) 分配收益資訊-公告日期 114年05月19日', 'https://mopsov.twse.com.tw/mops/web/ajax_t108sb19?step=2&firstin=1&TYPEK=all&SKIND=F&SEQ_NO=1&DATE1=20250519&COMP=A00016', None]
['2025-05-1

## LLM 可以做到嗎?
2025/5/16 11:00 測試當下截圖如下
![alt text](20250516最新訊息.png)

chatGPT

![alt text](20250516測試chatGPT上網.png)

Gemini

![alt text](20250516測試Gemini上網.png)

Grok

![alt text](20250516測試Grok上網.png)

Claude

![alt text](20250516測試Claude上網.png)

## 再問一次 chatGPT

![alt text](20250516測試chatGPT上網2.png)

他看不到了，因為他是 GET 爬蟲，他只能看到第一次送請求就能收到的內容

|LLM|最新資料時間|正確性|
|:-:|:-:|:-:|
|chatGPT|3/18|部分錯誤|
|Gemini|5/14|完全正確|
|Grok|5/15|完全錯誤|
|Claude|無法上網||

能上網的 LLM 都能解析出那些亂碼

有列出截至"日期"為止的，其實只是呼叫現在的日期(2025/5/16)，不是當場去看網頁

關於 LLM 搜尋引擎，會在之後詳談

## 課後練習
抓取台灣證券交易所最新訊息中，最近一個月的配息資訊

1. 公告標題中包含「分配收益資訊」
2. 挑選今天往前 30 天的資料（年份是民國要怎麼改成西元？如果今天是 2025 年 4 月 21 日，怎麼往前 30 天？）
3. 點進網址之後抓取配息金額（配息金額在不同的標的是否會出現在不同位置？）
4. 取得除息交易日並計算出最後買進日

## POST 請求
在 requests 爬蟲的 SOP 裡面，送請求的語法是 r = requests.get(url)，當我們不需要給網頁資料、直接跟網頁拿資料的時候，使用的是 get；當需要送一些資料，讓網頁伺服器決定要從資料庫撈什麼資料回傳的時候，使用的是 post。

例如要在台鐵時刻表( https://www.railway.gov.tw/tra-tip-web/tip/tip001/tip112/gobytime )取得今天從台北車站到台中車站的班次。

先在出發站和抵達站選好，然後打開開發人員工具的 Network，再按查詢，第一個跳出的就是你對網頁送的 POST 請求的資料。先看標頭(header)，可以在要求方法看到這是一個 POST 請求，而網址也跟原本的網址不一樣。把這個網址複製下來貼到 url，然後去看酬載(payload)，把裡面的每一個變數名稱和數值抄到程式碼裡，並整理成字典，結果如下面的程式碼：


In [6]:
url = 'https://www.railway.gov.tw/tra-tip-web/tip/tip001/tip112/querybytime'
data = {'startStation': '1000-臺北', 'endStation': '3300-臺中', 'transfer': 'ONE', 'rideDate': '2025/05/28',
        'startOrEndTime': 'true', 'startTime': '00:00', 'endTime': '23:59', 'trainTypeList': 'ALL',
        'queryClassification': 'NORMAL', 'query': '查詢'}
r = requests.post(url, headers=header, data=data)
soup = BeautifulSoup(r.text, 'html.parser')
soup

<!DOCTYPE html>

<html lang="zh-tw">
<head>
<meta charset="utf-8"/>
<!--電子分類檢索START-->
<!-- <meta http-equiv="Pragma" content="no-cache" /> -->
<!-- <meta http-equiv="Cache-Control" content="no-cache" /> -->
<meta content="國營臺灣鐵路股份有限公司" http-equiv="Expires">
<meta content="https://tip-tr4cdn.cdn.hinet.net/tra-tip-web/static/images/train.jpg" property="og:image"/>
<meta content="國營臺灣鐵路股份有限公司" property="og:title"/>
<!-- <meta name="_csrf" th:content="${_csrf.token}" /> -->
<!-- <meta name="_csrf_header" th:content="${_csrf.headerName}" /> -->
<meta content="TR WebSite - tw" name="DC.Title"/>
<meta content="國營臺灣鐵路股份有限公司" name="DC.Creator"/>
<meta content="TR WebSite - tw" name="DC.Subject"/>
<meta content="" name="DC.Description"/>
<meta content="國營臺灣鐵路股份有限公司" name="DC.Contributor"/>
<meta content="Text" name="DC.Type"/>
<meta content="text/html" name="DC.Format"/>
<meta content="國營臺灣鐵路股份有限公司" name="DC.Source"/>
<meta content="中文" name="DC.Language"/>
<meta content="" name="DC.coverage.t.

In [7]:
# 確實有想要的資料
soup.select('tbody tr[class="trip-column"] td:nth-child(2)')

[<td>05:24</td>,
 <td>05:46</td>,
 <td>06:02</td>,
 <td>07:00</td>,
 <td>07:04</td>,
 <td>07:45</td>,
 <td>08:00</td>,
 <td>08:03</td>,
 <td>08:28</td>,
 <td>09:00</td>,
 <td>09:05</td>,
 <td>10:00</td>,
 <td>10:20</td>,
 <td>11:00</td>,
 <td>11:05</td>,
 <td>11:58</td>,
 <td>12:03</td>,
 <td>12:25</td>,
 <td>12:48</td>,
 <td>12:58</td>,
 <td>13:28</td>,
 <td>13:47</td>,
 <td>14:00</td>,
 <td>15:00</td>,
 <td>15:10</td>,
 <td>15:40</td>,
 <td>15:50</td>,
 <td>16:00</td>,
 <td>16:40</td>,
 <td>17:00</td>,
 <td>17:28</td>,
 <td>17:37</td>,
 <td>18:00</td>,
 <td>18:24</td>,
 <td>18:43</td>,
 <td>18:59</td>,
 <td>19:30</td>,
 <td>20:20</td>,
 <td>21:00</td>]

可以觀察到，使用 post 的時候，多了一個 data 參數要填，這個 data 的內容就是我們剛剛抄下來的那些變數。其他部分都跟 get 請求一樣。現在可以看到 soup 內容就包含我們要的車次資料。只不過需要再想辦法定位出我們要的資料。

我們在用 python 寫函式的時候，會把一些變數設定預設值，在沒有資料輸入函式的時候，會自動填入這些預設值，網站伺服器也一樣。因此 POST 請求 data 裡面的變數可以只填一部分，但是哪些可以省略、哪些是必要的，必須要逐個刪掉測試才會知道。以台鐵時刻表來說，最精簡的變數如下：
```python
data = {'startStation': '1000-臺北', 'endStation': '3300-臺中', 'transfer': 'ONE', 'rideDate': '2025/05/16',
        'startTime': '00:00', 'endTime': '23:59', 'trainTypeList': 'ALL'}
```

In [10]:
url = 'https://www.railway.gov.tw/tra-tip-web/tip/tip001/tip112/querybytime'
data = {'startStation': '1000-臺北', 'endStation': '3300-臺中', 'transfer': 'ONE', 'rideDate': '2025/05/28',
        'startTime': '00:00', 'endTime': '23:59', 'trainTypeList': 'ALL'}
r = requests.post(url, headers=header, data=data)
soup = BeautifulSoup(r.text, 'html.parser')
soup.select('tbody tr[class="trip-column"] td:nth-child(2)')

[<td>05:24</td>,
 <td>05:46</td>,
 <td>06:02</td>,
 <td>07:00</td>,
 <td>07:04</td>,
 <td>07:45</td>,
 <td>08:00</td>,
 <td>08:03</td>,
 <td>08:28</td>,
 <td>09:00</td>,
 <td>09:05</td>,
 <td>10:00</td>,
 <td>10:20</td>,
 <td>11:00</td>,
 <td>11:05</td>,
 <td>11:58</td>,
 <td>12:03</td>,
 <td>12:25</td>,
 <td>12:48</td>,
 <td>12:58</td>,
 <td>13:28</td>,
 <td>13:47</td>,
 <td>14:00</td>,
 <td>15:00</td>,
 <td>15:10</td>,
 <td>15:40</td>,
 <td>15:50</td>,
 <td>16:00</td>,
 <td>16:40</td>,
 <td>17:00</td>,
 <td>17:28</td>,
 <td>17:37</td>,
 <td>18:00</td>,
 <td>18:24</td>,
 <td>18:43</td>,
 <td>18:59</td>,
 <td>19:30</td>,
 <td>20:20</td>,
 <td>21:00</td>]

### css selector 2
語法規則很多種，以爬蟲來說，爬比較新的網站，只要會 3 種就可以打天下。
1. 標籤  p # 抓標籤下的所有層的 p 標籤。
2. 標籤 > p # 抓標籤下一層的所有 p 標籤。
3. 標籤 p:nth-child(n) # 抓標籤下的第 n 個 p 標籤。從 1 開始

#### 標籤可以做限定
比不指定更能縮限範圍，在此只是以class這種屬性舉例，可以指定其他屬性。
1. 標籤[class="字串"] # 抓 class 為指定字串的所有標籤
2. 標籤[class*="字串"] # 抓 class 包含指定字串的所有標籤
3. 標籤[class^="字串"] # 抓 class 以指定字串開頭的所有標籤
4. 標籤[class$="字串"] # 抓 class 以指定字串結尾的所有標籤

組合起來會像是：標籤[class="字串"] p

In [ ]:
# 利用第一種把剛剛抓到的資料定位出來

data = {}
for row in soup.select('tbody tr[class="trip-column"]'):
    #因為要取全部所以取'td li'就好了 用text取純文字 然後把換行符號刪掉(用replace)
    train_number = row.select_one('td li').text.replace('\n', '')
    start_time = row.select('td')[1].text
    end_time = row.select('td')[2].text
    duration = row.select('td')[3].text
    data[train_number] = [train_number, start_time, end_time, duration]
    # break

data

{'區間快 2005(七堵→彰化)': ['區間快 2005(七堵→彰化)', '05:24', '07:55', '2 小時 31 分'],
 '區間 2133(基隆→嘉義)': ['區間 2133(基隆→嘉義)', '05:46', '09:27', '3 小時 41 分'],
 '自強(3000) 161(南港→臺東)': ['自強(3000) 161(南港→臺東)', '06:02', '08:03', '2 小時 1 分'],
 '自強 105(基隆→屏東)': ['自強 105(基隆→屏東)', '07:00', '09:18', '2 小時 18 分'],
 '區間 2143(基隆→嘉義)': ['區間 2143(基隆→嘉義)', '07:04', '10:20', '3 小時 16 分'],
 '普悠瑪 107(南港→潮州)': ['普悠瑪 107(南港→潮州)', '07:45', '09:21', '1 小時 36 分'],
 '自強(3000) 111(七堵→枋寮)': ['自強(3000) 111(七堵→枋寮)', '08:00', '09:49', '1 小時 49 分'],
 '區間 2153(基隆→嘉義)': ['區間 2153(基隆→嘉義)', '08:03', '11:28', '3 小時 25 分'],
 '自強 113(七堵→屏東)': ['自強 113(七堵→屏東)', '08:28', '10:39', '2 小時 11 分'],
 '自強 115(基隆→屏東)': ['自強 115(基隆→屏東)', '09:00', '11:18', '2 小時 18 分'],
 '區間 2163(基隆→嘉義)': ['區間 2163(基隆→嘉義)', '09:05', '12:21', '3 小時 16 分'],
 '自強(3000) 131(七堵→潮州)': ['自強(3000) 131(七堵→潮州)', '10:00', '11:49', '1 小時 49 分'],
 '自強(3000) 117(基隆→潮州)': ['自強(3000) 117(基隆→潮州)', '10:20', '12:24', '2 小時 4 分'],
 '自強 121(七堵→屏東)': ['自強 121(七堵→屏東)', '11:00', '13:15', '2

In [ ]:
# 利用第三種把剛剛抓到的資料定位出來

data = {}
for row in soup.select('tbody tr[class="trip-column"]'):
    #因為要取全部所以取'td li'就好了 用text取純文字 然後把換行符號刪掉(用replace)
    #select_one比select快
    train_number = row.select_one('td li').text.replace('\n', '')
    #nth-child裡面的數字是從1開始
    start_time = row.select_one('td:nth-child(2)').text
    end_time = row.select_one('td:nth-child(3)').text
    duration = row.select_one('td:nth-child(4)').text
    data[train_number] = [train_number, start_time, end_time, duration]
    # break

data

NameError: name 'soup' is not defined

## 使用API
API 的中文是應用程式介面，是網站的資料庫提供資料的一種管道。可以是網站自用，也可以提供給外面的人用。
+ 對網站來說，不會有爬蟲去佔網頁的頻寬；對爬蟲來說，有更方便取得資料的管道。雙贏。
+ 有些 API 需要申請且綁定帳號，這類的 API 對於網站來說是一種可以賣錢的服務，例如 OpenAI 和 google map。有些網站在設計上，前端會打 API 跟資料庫拿資料，我們可以在開發人員工具的 Network 裡面找到這樣的 API 網址，直接用這個網址跟資料庫拿資料，而不需要等網頁前端把東西都擺好之後才抓，速度會快很多。
+ 通常利用 API 爬蟲要用 requests 送請求，伺服器通常回傳 json 格式，所以要再用 json 轉成字典。


### google map API
具體用法請參考下面連結的講義，很詳細
https://ccclub-staging.s3.amazonaws.com/static/courses/resources/SLIDE/google-map-api.pdf

但是因為使用起來有限制，API 有點摳，所以通常建議學員直接爬 google map

## 3 個爬蟲模組比較
在爬蟲專案常用模組的時候，提到了 pandas、requests、selenium，這三個模組各有適合的使用情境。
### pandas
+ 主要用於資料分析，爬蟲功能僅限於抓表格
  1. 直接送檔案的用 read_csv，政府的開放資料大多可以這樣抓
  2. 表格在網頁裡的用 read_html，可以直接這樣抓的網頁越來越少了，但可搭配 requests 或 selenium 使用
+ 可以用的話，爬回來的資料就是可以用 pandas 繼續處理的格式，很方便。

### requests
+ 模擬瀏覽器的行為，主要用 get 和 post 抓資料。速度通常比 selenium 快，但有些網站會特別拖慢 requests 請求的回應的速度。
+ 如果網站有提供 API，可直接用 API 抓資料。
+ 如果網站本身從資料庫抓資料用 API，可直接去找到 API 網址，然後用 requests 抓資料。

### selenium
+ 本意是測試網頁的模組，被用來當作爬蟲
+ 模擬人類的行為，比 requests 更像人，但並不代表不會被抓到
+ 通常只要想抓的資料用 JavaScript 包起來，而且在 Network 找不到那包資料，就只能用 selenium（因為沒有實際開一個瀏覽器，缺的資料不會補）
+ 就算 selenium 很厲害，也不能肆無忌憚，凡是要登入會員的網站，你亂來就還是被 ban


模組的使用時機，以遊戲來比喻的話

pandas 像是屬性相剋技，剛好有剋到效果就很好，沒剋到就沒效。

requests 像是速攻技，出招速度快，但如果遇到比較硬的對手（防爬機制強），就容易沒效果。

selenium 像是大招，出招較慢，幾乎能破防（幾乎任何防爬機制都能突破或繞過）。


In [91]:
# 安裝模組
! pip install -v pandas==2.0.0 # 指定版本號
! pip install requests
! pip install beautifulsoup4
! pip install selenium

Using pip 23.0.1 from C:\Users\hchbi\anaconda3\envs\RegularTracking\Lib\site-packages\pip (python 3.11)


ERROR: Invalid requirement: '#'


## pandas 爬蟲：read_csv
在政府網站的資料開放區、行政院資料開放平台，通常有個地方可以下載 csv 檔，把下載的連結複製之後，貼到 url，再用 pandas 的 read_csv 讀就可以直接用了。以 NCC 的行動通信業務基地臺統計數為例

政府網站的資料開放區
https://www.ncc.gov.tw/chinese/opendata_item.aspx?menu_function_sn=208 # 檔案下載 → csv → 複製連結網址

政府資料開放平台
https://data.gov.tw/dataset/22592 # 檢視資料 → 資料下載網址 → 滑鼠右鍵 → 複製連結網址

In [1]:
import pandas as pd

url = 'https://www.ncc.gov.tw/chinese/files/opendata/%E8%A1%8C%E5%8B%95%E9%80%9A%E4%BF%A1%E6%A5%AD%E5%8B%99%E5%9F%BA%E5%9C%B0%E8%87%BA%E7%B5%B1%E8%A8%88%E6%95%B8(%E5%8D%80%E5%88%86%E7%B8%A3%E5%B8%82%E6%A5%AD%E8%80%85).csv'
#url = 'https://www.ncc.gov.tw/chinese/files/opendata/行動通信業務基地臺統計數(區分縣市業者).csv' # 這樣是不行的
df = pd.read_csv(url, header=0) # 這裡的 header 指的是要用第幾列資料當作表頭
df


ModuleNotFoundError: No module named 'pandas'

## pandas 爬蟲：read_html
看到網頁裡面有表格，就要想到可以用 read_html。read_html 同時有送請求和解析的功能，送請求的部分只能用 get 請求，似乎沒什麼偽裝，所以可以直接放網址的網頁越來越少。但只要先用 requests 取得網頁原始碼，再用 read_html 解析，就可以省去用 BeautifulSoup 解析和定位的功夫。
回到剛才台鐵時刻表的例子，原本的程式碼如下：

In [20]:
url = 'https://www.railway.gov.tw/tra-tip-web/tip/tip001/tip112/querybytime'
data = {'startStation': '1000-臺北', 'endStation': '3300-臺中', 'transfer': 'ONE', 'rideDate': '2025/05/28',
        'startTime': '00:00', 'endTime': '23:59', 'trainTypeList': 'ALL'}
r = requests.post(url, headers=header, data=data)
soup = BeautifulSoup(r.text, 'html.parser')

data = {}
for row in soup.select('tbody tr[class="trip-column"]'):
    train_number = row.select_one('td li').text.replace('\n', '')
    start_time = row.select_one('td:nth-child(2)').text
    end_time = row.select_one('td:nth-child(3)').text
    duration = row.select_one('td:nth-child(4)').text
    data[train_number] = [train_number, start_time, end_time, duration]

df = pd.DataFrame(data).transpose()
df

0      1      2          3
區間快 2005(七堵→彰化)          區間快 2005(七堵→彰化)  05:24  07:55  2 小時 31 分
區間 2133(基隆→嘉義)            區間 2133(基隆→嘉義)  05:46  09:27  3 小時 41 分
自強(3000) 161(南港→臺東)  自強(3000) 161(南港→臺東)  06:02  08:03   2 小時 1 分
自強 105(基隆→屏東)              自強 105(基隆→屏東)  07:00  09:18  2 小時 18 分
區間 2143(基隆→嘉義)            區間 2143(基隆→嘉義)  07:04  10:20  3 小時 16 分
普悠瑪 107(南港→潮州)            普悠瑪 107(南港→潮州)  07:45  09:21  1 小時 36 分
自強(3000) 111(七堵→枋寮)  自強(3000) 111(七堵→枋寮)  08:00  09:49  1 小時 49 分
區間 2153(基隆→嘉義)            區間 2153(基隆→嘉義)  08:03  11:28  3 小時 25 分
自強 113(七堵→屏東)              自強 113(七堵→屏東)  08:28  10:39  2 小時 11 分
自強 115(基隆→屏東)              自強 115(基隆→屏東)  09:00  11:18  2 小時 18 分
區間 2163(基隆→嘉義)            區間 2163(基隆→嘉義)  09:05  12:21  3 小時 16 分
自強(3000) 131(七堵→潮州)  自強(3000) 131(七堵→潮州)  10:00  11:49  1 小時 49 分
自強(3000) 117(基隆→潮州)  自強(3000) 117(基隆→潮州)  10:20  12:24   2 小時 4 分
自強 121(七堵→屏東)              自強 121(七堵→屏東)  11:00  13:15  2 小時 15 分
區間 2183(基隆→嘉義)            區間 2183(基隆→嘉義)  11:05  14:27  3 小時 22 分
自強 123(七堵→屏東)              自強 123(七堵→屏東)  11:58  14:13  2 小時 15 分
區間 2193(基隆→嘉義)            區間 2193(基隆→嘉義)  12:03  15:29  3 小時 26 分
普悠瑪 271(花蓮→彰化)            普悠瑪 271(花蓮→彰化)  12:25  14:18  1 小時 53 分
區間 2203(基隆→嘉義)            區間 2203(基隆→嘉義)  12:48  16:08  3 小時 20 分
自強 125(七堵→屏東)              自強 125(七堵→屏東)  12:58  15:19  2 小時 21 分
普悠瑪 165(南港→臺東)            普悠瑪 165(南港→臺東)  13:28  15:08  1 小時 40 分
區間 2213(基隆→嘉義)            區間 2213(基隆→嘉義)  13:47  17:10  3 小時 23 分
自強 129(基隆→枋寮)              自強 129(基隆→枋寮)  14:00  16:11  2 小時 11 分
自強 135(七堵→屏東)              自強 135(七堵→屏東)  15:00  17:14  2 小時 14 分
區間 2223(基隆→嘉義)            區間 2223(基隆→嘉義)  15:10  18:39  3 小時 29 分
普悠瑪 137(南港→潮州)            普悠瑪 137(南港→潮州)  15:40  17:20  1 小時 40 分
區間 2233(瑞芳→嘉義)            區間 2233(瑞芳→嘉義)  15:50  19:20  3 小時 30 分
自強 175(花蓮→潮州)              自強 175(花蓮→潮州)  16:00  18:15  2 小時 15 分
普悠瑪 273(花蓮→彰化)            普悠瑪 273(花蓮→彰化)  16:40  18:20  1 小時 40 分
自強 139(七堵→潮州)              自強 139(七堵→潮州)  17:00  19:15  2 小時 15 分
自強 141(七堵→潮州)              自強 141(七堵→潮州)  17:28  19:52  2 小時 24 分
自強 177(花蓮→斗南)              自強 177(花蓮→斗南)  17:37  20:18  2 小時 41 分
自強(3000) 143(七堵→潮州)  自強(3000) 143(七堵→潮州)  18:00  19:47  1 小時 47 分
自強(3000) 477(臺東→員林)  自強(3000) 477(臺東→員林)  18:24  20:30   2 小時 6 分
區間快 2037(七堵→臺中)          區間快 2037(七堵→臺中)  18:43  21:28  2 小時 45 分
自強 149(基隆→潮州)              自強 149(基隆→潮州)  18:59  21:14  2 小時 15 分
自強 181(花蓮→屏東)              自強 181(花蓮→屏東)  19:30  21:42  2 小時 12 分
自強(3000) 283(壽豐→斗六)  自強(3000) 283(壽豐→斗六)  20:20  22:15  1 小時 55 分
自強 155(七堵→彰化)              自強 155(七堵→彰化)  21:00  23:18  2 小時 18 分

In [1]:
# 還需要定位出自己想要的資料在哪，但如果改成：
url = 'https://www.railway.gov.tw/tra-tip-web/tip/tip001/tip112/querybytime'
data = {'startStation': '1000-臺北', 'endStation': '3300-臺中', 'transfer': 'ONE', 'rideDate': '2025/05/28',
        'startTime': '00:00', 'endTime': '23:59', 'trainTypeList': 'ALL'}
r = requests.post(url, headers=header, data=data)
df = pd.read_html(r.text)[0] # read_html會把所有表格都抓出來，想要的資料在第一個表格
df = df[df['出發時間'].str.match('\d{2}:\d{2}(?!.)')] # 由於台鐵時刻表的表格特別髒亂，需要做一些篩選。不懂沒關係，示範用而已
df = df[['車種車次 (始發站 → 終點站)', '出發時間', '抵達時間', '行駛時間']]
df

# 整個清單就直接整理好了

NameError: name 'requests' is not defined

## LLM 做得到嗎?

chatGPT

![alt text](20250516測試chatGPT查火車.png)

Gemini

![alt text](20250516測試Gemini查火車.png)

Grok

![alt text](20250516測試Grok查火車.png)

LLM 無法做 POST 爬蟲

## 下載 driver

以 chromedriver 為例，chromedriver 的版本需要和瀏覽器版本一致（主版本號一致就好，例如目前是136）才可以使用。但如果你剛剛是用 pip 安裝 selenium，你的版本號會是 4.32，你可以直接跳過這個步驟，如果你在以前已經安裝過比較舊的 selenium 版本，建議更新。如果你在以前已經抓過比較舊的 driver，建議刪掉。

在 4.6 之後，selenium 會自動偵測有沒有 driver，如果沒有的話就會自動下載到

(Windows) C:\Users\使用者\.cache\selenium\chromedriver\win64

(Mac) $HOME\.cache\selenium

(Linux) $HOME\.cache\selenium

因此使用者再也不用管 driver 版本跟瀏覽器版本是否有一致。缺點是他不會刪掉舊的 driver，所以有時候還是要來刪一下舊的 driver。

# 下集預告
+ selenium 爬蟲
+ 進度條
+ 下載檔案
+ 等待連線
+ 例行爬蟲
+ 驗證碼
+ 爬蟲的種類
+ 爬蟲的智慧財產權問題
+ 好用模組推薦
+ 其他學習資源